# Экзамен по курсу "Аналитика данных на Python"

Этот тест проверит Ваши навыки работы с таблицами данных с помощью библиотек pandas и numpy. Задания делятся на простые (⭐️), средние (⭐️⭐️) и сложные (⭐️⭐️⭐️). Решение простых заданий, как правило, требует одной-двух операций с таблицами, тогда как для более сложных может потребоваться несколько последовательных преобразований данных.

##  Предыстория

Сегодня Ваш первый день работы в крупном интернет-магазине, который продаёт товары с доставкой по всему миру. Ваша первая задача – проанализировать базу данных покупок, совершённых в магазине за последние несколько лет. База содержит информацию об отдельных транзакциях, про каждую из которых известны номер инвойса (InvoiceNo), дата инвойса(InvoiceDate), код товара (StockCode), описание товара(Description), количество товара в транзакции (Quantity), стоимость единицы товара (UnitPrice), код покупателя (CustomerID), страна покупателя (Country).

База выгружена для Вас в формате CSV: файл [```online-retail.csv```](https://drive.google.com/file/d/1O1oJtpEu-u6s6xTu7seQfnNcQjYCt0GF/view?usp=sharing)

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## ⭐️ Вопрос 1

Загрузите данные из файла online-retail.csv в переменную типа pandas DataFrame

Подсказка: Используйте функцию из библиотеки pandas

Какой символ-разделитель используется в этом файле?

* Запятая ","
* Двоеточие ":"
* Точка с запятой ";"
* Символ табуляции "tab"

#### Решение

In [5]:
df = pd.read_csv('online-retail.csv', parse_dates=['InvoiceDate'], sep=';')
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541906    581587     23254         CHILDRENS CUTLERY DOLLY GIRL         4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138         BAKING SET 9 PIECE RETROSPOT         3   
541909    581588     22380                    TOY TIDY SPACEBOY         9   
541910    581588     22381               TOY TIDY PINK POLKADOT         9   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  
541909 2011-12-09 12:52:00       2.10     12594.0           Italy  
541910 2011-12-09 12:52:00       2.10     12594.0           Italy  

[541911 rows x 8 columns]

Ответ: точка с запятой.

## ⭐️ Вопрос 2

Сколько строк в полученной таблице (не считая заголовков столбцов)?

#### Решение

In [6]:
# Ответ, 541 911 тысяч строк в текущем дата фрейме.
df.shape[0]

541911

## ⭐️ Вопрос 3

Сколько столбцов в полученной таблице (не считая индекса)?

#### Решение

In [7]:
# Ответ, 8 столбцев
df.shape[1]

8

## ⭐️ Вопрос 4

Как называется столбец с самым коротким названием?

#### Решение

In [8]:
min(df.columns, key=len)

'Country'

## ⭐️ Вопрос 5

В скольких столбцах встречаются пропущенные значения? (ответ - целое число)

#### Решение

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541911 entries, 0 to 541910
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541911 non-null  object        
 1   StockCode    541911 non-null  object        
 2   Description  540456 non-null  object        
 3   Quantity     541911 non-null  int64         
 4   InvoiceDate  541911 non-null  datetime64[ns]
 5   UnitPrice    541911 non-null  float64       
 6   CustomerID   406831 non-null  float64       
 7   Country      541911 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [11]:
df.isna().sum()

InvoiceNo           0
StockCode           0
Description      1455
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

В двух столбцах пропущенны данные, в столбце Description и CustomerID

## ⭐️ Вопрос 6

Сколько пропущенных значений в столбце CustomerID? (ответ - целое число)

#### Решение

In [12]:
df.CustomerID.isna().sum()

135080

Ответ: 135 080

## ⭐️⭐️ Вопрос 7

Посмотрим, данные за какой исторический период у нас есть. 

Данные за какой самый ранний и за какой самый поздний годы содержатся в датасете? В ответе укажите два целых числа через запятую.

#### Решение

In [13]:
df.InvoiceDate.min().year

2010

In [15]:
df.InvoiceDate.max().year

2011

Ответ: 2010, 2011

## ⭐️ Вопрос 8

Каковы минимальная и максимальная цена товаров (UnitPrice)? Перечислите через запятую. Цена в этом задании может принимать отрицательные значения.

#### Решение

In [17]:
df.UnitPrice.min()

-11062.06

In [18]:
df.UnitPrice.max()

38970.0

## ⭐️ Вопрос 9

В таблице оказались товары с отрицательными ценами! Это явно какая-то ошибка. Какое описание (Description) у таких транзакций? Перечислите все варианты через запятую, отсортировав строки по алфавиту.

#### Решение

In [19]:
', '.join(df[df.UnitPrice < 0].Description.sort_values().tolist())

'Adjust bad debt, Adjust bad debt'

Ответ: 'Adjust bad debt, Adjust bad debt'

## ⭐️ Вопрос 10

Поищем ещё возможные проблемы с данными. Как насчёт товаров с нулевыми ценами? 

Сколько в таблице транзакций с нулевой ценой? А с пропусками на месте цены? 

Перечислите два целых числа через запятую.

#### Решение

In [21]:
df.UnitPrice.eq(0).sum()

2515

In [22]:
df.UnitPrice.isna().sum()

0

Ответ: 2515, 0

## ⭐️⭐️ Вопрос 11

Для дальнейшего анализа поведения покупателей нам понадобится набор данных, в которых у каждой транзакции корректно указана цена, количество единиц товара (Quantity) и id покупателя (CustomerID). Удалите из таблицы все строки, в которых цена не превосходит 0 или пропущена, или количество единиц товара не превосходит 0 или пропущено, или в которых пропущен id покупателя. 

Сколько строк осталось?

#### Решение

In [25]:
indexes_for_drop = df[(df.UnitPrice <= 0) | df.UnitPrice.isna() |
                      (df.Quantity <= 0) | df.Quantity.isna() |
                       df.CustomerID.isna()].index
df.drop(indexes_for_drop, inplace=True)
df.shape[0]

397886

Ответ: 397 886

## Внимание! 

### Везде далее мы работаем с очищенной таблицей, полученной в Вопросе 11.

## ⭐️⭐️ Вопрос 12

В таблице для каждой транзакции указаны цена за единицу товара (UnitPrice) и количество единиц товара (Quantity). Вычислите для каждой транзакции её **полную стоимость** и сохраните в новом столбце Price, который добавьте в таблицу. 

Каковы минимальная и максимальная полная стоимость транзакций? Перечислите через запятую, округлив до целых чисел.

#### Решение

In [26]:
df.loc[:, 'Price'] = df.UnitPrice * df.Quantity

In [27]:
round(df.Price.min())

0

In [29]:
round(df.Price.max())

168470

Ответ: минимальное 0, максимальное 168 470

## ⭐️⭐️ Вопрос 13

В нашей базе данных каждая покупка представлена одной или несколькими транзакциями. Покупка однозначно определяется своим номером инвойса (InvoiceNo). Транзакции, относящиеся к одной покупке, имеют один и тот же InvoiceNo. 

Стоимость всей покупки равна сумме полных стоимостей транзакций, входящих в неё. Найдите стоимости трёх самых дорогих покупок. Перечислите через запятую в порядке убывания, округлив до целых чисел.

#### Решение

In [30]:
df.groupby('InvoiceNo').Price.sum().sort_values(ascending=False).head(3)

InvoiceNo
581483    168469.6
541431     77183.6
556444     38970.0
Name: Price, dtype: float64

In [31]:
highest_prices = df.groupby('InvoiceNo').agg({'Price':'sum'})\
                   .sort_values('Price', ascending=False)\
                   .Price.head(3)
for price in highest_prices:
    print(round(price), end=', ')

168470, 77184, 38970, 

Ответ: 168470, 77184, 38970

## ⭐️⭐️ Вопрос 14

Какой товар составил наибольшую выручку? В ответе укажите описание товара (дословно строку из соответствующего поля в столбце Description).

#### Решение

In [32]:
print(df.groupby('Description').agg({'Price':'sum'}).sort_values('Price', ascending=False).index[0])

PAPER CRAFT , LITTLE BIRDIE


Ответ: PAPER CRAFT , LITTLE BIRDIE

## ⭐️⭐️ Вопрос 15

Создайте новую таблицу purchases, в которой каждая строка будет соответствовать отдельной покупке, со столбцами 
InvoiceNo, InvoiceDate, Price, CustomerID, Country.

Поскольку дата InvoiceDate может быть отличаться для разных транзакций внутри одной покупки, при группировке возьмите самую раннюю из дат. Также отсортируйте её по датам по возрастанию.

Сколько получилось строк?

#### Решение

In [33]:
purchases = df[['InvoiceNo', 'InvoiceDate', 'Price', 'CustomerID', 'Country']]

In [34]:
purchases_grouped = purchases.groupby('InvoiceNo').agg({'InvoiceDate':'min',
                                                        'Price':'sum',
                                                        'CustomerID':'min',
                                                        'Country':'min'})
purchases_grouped

InvoiceDate   Price  CustomerID         Country
InvoiceNo                                                        
536365    2010-12-01 08:26:00  139.12     17850.0  United Kingdom
536366    2010-12-01 08:28:00   22.20     17850.0  United Kingdom
536367    2010-12-01 08:34:00  278.73     13047.0  United Kingdom
536368    2010-12-01 08:34:00   70.05     13047.0  United Kingdom
536369    2010-12-01 08:35:00   17.85     13047.0  United Kingdom
...                       ...     ...         ...             ...
581584    2011-12-09 12:25:00  140.64     13777.0  United Kingdom
581585    2011-12-09 12:31:00  329.05     15804.0  United Kingdom
581586    2011-12-09 12:49:00  339.20     13113.0  United Kingdom
581587    2011-12-09 12:50:00  249.45     12680.0          France
581588    2011-12-09 12:52:00   37.80     12594.0           Italy

[18533 rows x 4 columns]

In [36]:
purchases_grouped.shape[0]

18533

Ответ:
18533

## ⭐️⭐️ Вопрос 16

Исследуем, растут или убывают покупки наших клиентов после их первой покупки на сайте.
Найдите среднюю цену первых покупок клиентов: для каждого клиента возьмите его первую покупку и усредните эти значения. Найдите среднюю цену покупок в целом. Округлите эти числа до целых и перечислите в этом порядке через запятую.

#### Решение

In [38]:
purchases.groupby(['CustomerID', 'InvoiceDate'])\
         .agg({'Price':'sum'})\
         .reset_index()\
         .drop_duplicates('CustomerID')\
         .Price.mean()\
         .__round__()

425

In [39]:
purchases.groupby(['InvoiceNo'])\
         .agg({'Price':'sum'})\
         .Price\
         .mean()\
         .__round__()

481

Ответ: 425 и 481

## ⭐️⭐️⭐️ Вопрос 17

В какой день недели было наибольшее число покупок? В ответе укажите русское название дня недели, начинающееся с заглавной буквы.


Подсказка: 
* Преобразуйте тип данных в столбце InvoiceDate таблицы purchases из строк в datetime. 
* Для каждой покупки вычислите день недели, в который она была совершена. Сохраните в новый столбец.
* Сгруппируйте таблицу по дням недели.

#### Решение

In [40]:
purchases = purchases.assign(DayName=purchases.InvoiceDate.dt.day_name(locale='Russian').values)

In [41]:
purchases.groupby('DayName').agg({'InvoiceNo':'nunique'})\
                            .sort_values('InvoiceNo', ascending=False)\
                            .head(1)

InvoiceNo
DayName           
Четверг       4032

## ⭐️⭐️⭐️ Вопрос 18

В какой год и месяц выручка была максимальной?

Подсказка: 
* Преобразуйте тип данных в столбце InvoiceDate таблицы purchases из строк в datetime. 
* Для каждой покупки вычислите год и месяц, в которые она была совершена. Сохраните в новые столбцы.
* Сгруппируйте таблицу по новым столбцам

В ответе укажите два целых числа через запятую: год, месяц

#### Решение

In [42]:
purchases = purchases.assign(Year=purchases.InvoiceDate.dt.year.values,
                             Month=purchases.InvoiceDate.dt.month.values)

In [43]:
grouped_by_purchases = purchases.groupby('InvoiceNo').agg({'Price':'sum', 'Year':'min', 'Month':'min'})

In [44]:
purchases_by_years = grouped_by_purchases.groupby('Year').agg({'Price':'sum'})
purchases_by_years[purchases_by_years.Price.eq(purchases_by_years.Price.max())]

Price
Year             
2011  8338731.814

In [45]:
purchases_by_months = grouped_by_purchases.groupby('Month').agg({'Price':'sum'})
purchases_by_months[purchases_by_months.Price.eq(purchases_by_months.Price.max())]

Price
Month            
11     1161817.38

Ответ: 2011, 11

## ⭐️⭐️⭐️ Вопрос 19

Магазин продаёт товары покупателям из разных стран (Country). В какой стране был наибольший процентный рост месячных продаж, если сравнить март 2011 и сентябрь 2011? Сколько процентов составил этот рост? В расчёт брать только страны, в которых были ненулевые продажи в обоих этих месяцах. В ответе укажите через запятую название страны и целое число (процентный рост, округлённый до целого числа).

#### Решение

In [46]:
march_sept_purchases = purchases[purchases.Year.eq(2011) &
                                 (purchases.Month.eq(3) | purchases.Month.eq(9))]

In [47]:
march_sept_purchases

InvoiceNo         InvoiceDate  Price  CustomerID         Country  \
105335    545220 2011-03-01 08:30:00  15.90     14620.0  United Kingdom   
105336    545220 2011-03-01 08:30:00  15.90     14620.0  United Kingdom   
105337    545220 2011-03-01 08:30:00  19.80     14620.0  United Kingdom   
105338    545220 2011-03-01 08:30:00  14.85     14620.0  United Kingdom   
105339    545220 2011-03-01 08:30:00  15.00     14620.0  United Kingdom   
...          ...                 ...    ...         ...             ...   
370675    569148 2011-09-30 15:52:00  17.00     16613.0  United Kingdom   
370676    569148 2011-09-30 15:52:00  17.00     16613.0  United Kingdom   
370677    569148 2011-09-30 15:52:00  25.20     16613.0  United Kingdom   
370678    569148 2011-09-30 15:52:00  25.20     16613.0  United Kingdom   
370679    569148 2011-09-30 15:52:00  10.20     16613.0  United Kingdom   

        DayName  Year  Month  
105335  Вторник  2011      3  
105336  Вторник  2011      3  
105337  Вторник  2011      3  
105338  Вторник  2011      3  
105339  Вторник  2011      3  
...         ...   ...    ...  
370675  Пятница  2011      9  
370676  Пятница  2011      9  
370677  Пятница  2011      9  
370678  Пятница  2011      9  
370679  Пятница  2011      9  

[67203 rows x 8 columns]

In [49]:
# сгруппируем по странам, годам и месяцам просуммировав цены продаж
pct_increase = march_sept_purchases.groupby(['Country', 'Year', 'Month'])\
                                   .agg({'Price':'sum'})\
                                   .unstack(2)\
                                   .dropna()

In [50]:
# рассчитаем рост продаж в процентах для каждой страны и отсортируем по убыванию
pct_increase = pct_increase.assign(PctIncrease=((pct_increase.Price[9] - pct_increase.Price[3])/pct_increase.Price[3] * 100)
                                                 .round()
                                                 .astype('int'))\
                                                 .sort_values('PctIncrease', ascending=False)

In [51]:
pct_increase.index[0][0]

'Norway'

In [52]:
pct_increase.PctIncrease[0]

586

Ответ: Норвегия, 586

## ⭐️⭐️⭐️ Вопрос 20

Большинство клиентов все свои покупки делают из одной и той же страны. Выясним, однако, насколько велика доля путешественников среди клиентов. 
Сколько клиентов сделали покупки по крайней мере из двух разных стран? (ответ - целое число)

#### Решение

In [53]:
traveling_byers = purchases.groupby('CustomerID').agg({'Country':'unique'})
traveling_byers = traveling_byers.Country.apply(pd.Series).dropna()
traveling_byers

0            1
CustomerID                          
12370.0          Cyprus      Austria
12394.0         Belgium      Denmark
12417.0         Belgium        Spain
12422.0       Australia  Switzerland
12429.0         Denmark      Austria
12431.0       Australia      Belgium
12455.0          Cyprus        Spain
12457.0     Switzerland       Cyprus

In [54]:
traveling_byers.index.size

8

Ответ: 8

## ⭐️⭐️⭐️ Вопрос 21

Мы запускаем в Италии рекомендательную систему "С этим товаром часто покупают...", и для этого хотим узнать, какие различные товары чаще всего встречаются в одной покупке из этой страны. Определите, какая пара различных товаров чаще всего встречается в различных покупках с ```Country=='Italy'```, и в скольких покупках это происходит. Одинаковые товары или нет, проверяйте по равенству поля Description. 

(ответ: название (Description) первого товара, название (Description) второго товара, целое число)

#### Решение

In [55]:
# оставим только покупки совершённые в Италии
Italy_purchases = df[df.Country.eq('Italy')]

In [56]:
# группировка транзакций в покупки и получение описания товара Description списком
Italy_purchases = Italy_purchases.groupby('InvoiceNo').agg({'Description':'unique'})

In [58]:
from itertools import combinations
from collections import defaultdict

# перебор всех комбинаций пар товаров и запись результата в словарь
purchases_dict = defaultdict(int)
for purchase in Italy_purchases.Description:
    for items in combinations(purchase, 2):
        purchases_dict[tuple(sorted(items))] += 1

In [59]:
# сортировка словаря по значениям по убыванию и выбор наибольшего
sorted(purchases_dict.items(), key=lambda x: x[1], reverse=True)[0]

(('TOY TIDY PINK POLKADOT', 'TOY TIDY SPACEBOY'), 6)

In [60]:
purchases_dict['TOY TIDY PINK POLKADOT', 'TOY TIDY SPACEBOY']

6

Ответ:
TOY TIDY PINK POLKADOT, TOY TIDY SPACEBOY, 6